cli/src/commands/touch.rs

```diff
@@ -0,0 +1,173 @@
+        workspace_command.parse_revset(ui, &RevisionArg::AT)?
+    }
+    .evaluate_to_commit_ids()?
+    .try_collect()?;
+    if commit_ids.is_empty() {
+        writeln!(ui.status(), "No revisions to touch.")?;
+        return Ok(());
+    }
+    workspace_command.check_rewritable(commit_ids.iter())?;
+
+    let mut tx = workspace_command.start_transaction();
+    let tx_description = match commit_ids.as_slice() {
+        [] => unreachable!(),
+        [commit] => format!("touch commit {}", commit.hex()),
+        [first_commit, remaining_commits @ ..] => {
+            format!(
+                "touch commit {} and {} more",
+                first_commit.hex(),
+                remaining_commits.len()
+            )
+        }
+    };
+
+    let mut num_touched = 0;
+    let mut num_reparented = 0;
+    let commit_ids_set: HashSet<_> = commit_ids.iter().cloned().collect();
+    let mut touched: Vec<(CommitId, CommitId)> = Vec::new();
+    // Even though `MutableRepo::rewrite_commit` and
+    // `MutableRepo::rebase_descendants` can handle rewriting of a commit even
+    // if it is a descendant of another commit being rewritten, using
+    // `MutableRepo::transform_descendants` prevents us from rewriting the same
+    // commit multiple times, and adding additional entries in the predecessor
+    // chain.
+    tx.repo_mut()
+        .transform_descendants(commit_ids, async |rewriter| {
+            let old_commit_id = rewriter.old_commit().id().clone();
+            let mut commit_builder = rewriter.reparent();
+            if commit_ids_set.contains(&old_commit_id) {
+                let mut new_author = commit_builder.author().clone();
+                if let Some((name, email)) = args.author.clone() {
+                    new_author.name = name;
+                    new_author.email = email;
+                } else if args.update_author {
+                    new_author.name = commit_builder.committer().name.clone();
+                    new_author.email = commit_builder.committer().email.clone();
+                }
+                if args.update_author_timestamp {
+                    new_author.timestamp = commit_builder.committer().timestamp;
+                }
+                commit_builder = commit_builder.set_author(new_author);
+
+                if args.update_change_id {
+                    commit_builder = commit_builder.generate_new_change_id();
+                }
+
+                let new_commit = commit_builder.write()?;
+                touched.push((old_commit_id, new_commit.id().clone()));
+                num_touched += 1;
+            } else {
+                commit_builder.write()?;
+                num_reparented += 1;
+            }
+            Ok(())
+        })?;
+    if args.update_change_id {
+        for (old, new) in touched {
+            tx.repo_mut().set_rewritten_commit(old, new);
+        }
+    }
+    if num_touched > 1 {
+        writeln!(ui.status(), "Updated {num_touched} commits")?;
```

Comment:
nit: maybe you mean `num_touched > 0`?

Reply:
Right. Perhaps, `jj describe` is/was the command to update a single commit, so `num_described > 1` is unusual. I don't think that would apply to `jj touch`. I think it's also good to print a list of updated commits like `jj sign`.